In this project, I will be using 
-nuniue() and unique()

-value_counts()

-Filtering

-groupby()

-getgroup()

-isnull() & notnull()

-rename()

-mean()

-std()

-var()

-str.contains()

-And(&) Operator

-Or(!) operator

-min() and max()


##### Weather Dataset
Weather dataset is a time-series dataset with per-hour information abbout the weather conditions at a particular location .It records temperature, dew point temperature ,Relative humidity ,Wind speed, Visibility, Pressure and conditions .

In [3]:
import pandas as pd 
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [6]:
df=pd.read_csv('Weather Dataset.csv')
df

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0        1/1/2012 0:00    -1.8              -3.9         86                4   
1        1/1/2012 1:00    -1.8              -3.7         87                4   
2        1/1/2012 2:00    -1.8              -3.4         89                7   
3        1/1/2012 3:00    -1.5              -3.2         88                6   
4        1/1/2012 4:00    -1.5              -3.3         88                7   
...                ...     ...               ...        ...              ...   
8779  12/31/2012 19:00     0.1              -2.7         81               30   
8780  12/31/2012 20:00     0.2              -2.4         83               24   
8781  12/31/2012 21:00    -0.5              -1.5         93               28   
8782  12/31/2012 22:00    -0.2              -1.8         89               28   
8783  12/31/2012 23:00     0.0              -2.1         86               30   

      Visibility_km  Press_kPa               Weather  
0               8.0     101.24                   Fog  
1               8.0     101.24                   Fog  
2               4.0     101.26  Freezing Drizzle,Fog  
3               4.0     101.27  Freezing Drizzle,Fog  
4               4.8     101.23                   Fog  
...             ...        ...                   ...  
8779            9.7     100.13                  Snow  
8780            9.7     100.03                  Snow  
8781            4.8      99.95                  Snow  
8782            9.7      99.91                  Snow  
8783           11.3      99.89                  Snow  

[8784 rows x 8 columns]

In [7]:
df.shape

(8784, 8)

In [8]:
df.value_counts()

Date/Time        Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  Visibility_km  Press_kPa  Weather     
1/1/2012 0:00    -1.8    -3.9              86         4                8.0            101.24     Fog             1
6/1/2012 12:00    19.3    3.3              35         20               48.3           101.32     Cloudy          1
5/9/2012 7:00     14.3    12.5             89         15               4.8            100.12     Fog             1
5/9/2012 8:00     14.3    12.3             88         17               6.4            100.12     Fog             1
5/9/2012 9:00     14.0    12.3             89         9                4.0            100.10     Drizzle,Fog     1
                                                                                                                ..
12/8/2012 3:00    2.1    -1.5              77         6                25.0           101.18     Cloudy          1
12/8/2012 2:00    2.0    -1.9              75         7                25.0          

In [9]:
df.info

<bound method DataFrame.info of              Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0        1/1/2012 0:00    -1.8              -3.9         86                4   
1        1/1/2012 1:00    -1.8              -3.7         87                4   
2        1/1/2012 2:00    -1.8              -3.4         89                7   
3        1/1/2012 3:00    -1.5              -3.2         88                6   
4        1/1/2012 4:00    -1.5              -3.3         88                7   
...                ...     ...               ...        ...              ...   
8779  12/31/2012 19:00     0.1              -2.7         81               30   
8780  12/31/2012 20:00     0.2              -2.4         83               24   
8781  12/31/2012 21:00    -0.5              -1.5         93               28   
8782  12/31/2012 22:00    -0.2              -1.8         89               28   
8783  12/31/2012 23:00     0.0              -2.1         86               30   

      V

In [10]:
df.head()

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0  1/1/2012 0:00    -1.8              -3.9         86                4   
1  1/1/2012 1:00    -1.8              -3.7         87                4   
2  1/1/2012 2:00    -1.8              -3.4         89                7   
3  1/1/2012 3:00    -1.5              -3.2         88                6   
4  1/1/2012 4:00    -1.5              -3.3         88                7   

   Visibility_km  Press_kPa               Weather  
0            8.0     101.24                   Fog  
1            8.0     101.24                   Fog  
2            4.0     101.26  Freezing Drizzle,Fog  
3            4.0     101.27  Freezing Drizzle,Fog  
4            4.8     101.23                   Fog

In [11]:
df.tail()

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
8779  12/31/2012 19:00     0.1              -2.7         81               30   
8780  12/31/2012 20:00     0.2              -2.4         83               24   
8781  12/31/2012 21:00    -0.5              -1.5         93               28   
8782  12/31/2012 22:00    -0.2              -1.8         89               28   
8783  12/31/2012 23:00     0.0              -2.1         86               30   

      Visibility_km  Press_kPa Weather  
8779            9.7     100.13    Snow  
8780            9.7     100.03    Snow  
8781            4.8      99.95    Snow  
8782            9.7      99.91    Snow  
8783           11.3      99.89    Snow

In [12]:
df.describe()

Temp_C  Dew Point Temp_C    Rel Hum_%  Wind Speed_km/h  \
count  8784.000000       8784.000000  8784.000000      8784.000000   
mean      8.798144          2.555294    67.431694        14.945469   
std      11.687883         10.883072    16.918881         8.688696   
min     -23.300000        -28.500000    18.000000         0.000000   
25%       0.100000         -5.900000    56.000000         9.000000   
50%       9.300000          3.300000    68.000000        13.000000   
75%      18.800000         11.800000    81.000000        20.000000   
max      33.000000         24.400000   100.000000        83.000000   

       Visibility_km    Press_kPa  
count    8784.000000  8784.000000  
mean       27.664447   101.051623  
std        12.622688     0.844005  
min         0.200000    97.520000  
25%        24.100000   100.560000  
50%        25.000000   101.070000  
75%        25.000000   101.590000  
max        48.300000   103.650000

In [13]:
df.index

RangeIndex(start=0, stop=8784, step=1)

In [14]:
df.columns

Index(['Date/Time', 'Temp_C', 'Dew Point Temp_C', 'Rel Hum_%',
       'Wind Speed_km/h', 'Visibility_km', 'Press_kPa', 'Weather'],
      dtype='object')

In [15]:
df.dtypes

Date/Time            object
Temp_C              float64
Dew Point Temp_C    float64
Rel Hum_%             int64
Wind Speed_km/h       int64
Visibility_km       float64
Press_kPa           float64
Weather              object
dtype: object

In [18]:
df['Weather'].unique

<bound method Series.unique of 0                        Fog
1                        Fog
2       Freezing Drizzle,Fog
3       Freezing Drizzle,Fog
4                        Fog
                ...         
8779                    Snow
8780                    Snow
8781                    Snow
8782                    Snow
8783                    Snow
Name: Weather, Length: 8784, dtype: object>

In [19]:
df['Weather'].value_counts()

Weather
Mainly Clear                               2106
Mostly Cloudy                              2069
Cloudy                                     1728
Clear                                      1326
Snow                                        390
Rain                                        306
Rain Showers                                188
Fog                                         150
Rain,Fog                                    116
Drizzle,Fog                                  80
Snow Showers                                 60
Drizzle                                      41
Snow,Fog                                     37
Snow,Blowing Snow                            19
Rain,Snow                                    18
Thunderstorms,Rain Showers                   16
Haze                                         16
Drizzle,Snow,Fog                             15
Freezing Rain                                14
Freezing Drizzle,Snow                        11
Freezing Drizzle                

In [20]:
df['Weather'].nunique()

50

In [22]:
df.count()

Date/Time           8784
Temp_C              8784
Dew Point Temp_C    8784
Rel Hum_%           8784
Wind Speed_km/h     8784
Visibility_km       8784
Press_kPa           8784
Weather             8784
dtype: int64

In [24]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8784 entries, 0 to 8783
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Date/Time         8784 non-null   object 
 1   Temp_C            8784 non-null   float64
 2   Dew Point Temp_C  8784 non-null   float64
 3   Rel Hum_%         8784 non-null   int64  
 4   Wind Speed_km/h   8784 non-null   int64  
 5   Visibility_km     8784 non-null   float64
 6   Press_kPa         8784 non-null   float64
 7   Weather           8784 non-null   object 
dtypes: float64(4), int64(2), object(2)
memory usage: 549.1+ KB


##### Find the unique 'Wind Speed value' in the dataset

In [26]:
df['Wind Speed_km/h'].value_counts()


Wind Speed_km/h
9     830
11    791
13    735
15    719
7     677
17    666
19    616
6     609
20    496
4     474
22    439
24    374
0     309
26    242
28    205
30    161
32    139
33     85
35     53
37     45
39     24
41     22
44     14
43     13
48     13
46     11
52      7
57      5
50      4
2       2
83      1
70      1
63      1
54      1
Name: count, dtype: int64

In [27]:
df['Wind Speed_km/h'].nunique()

34

In [29]:
df['Wind Speed_km/h'].unique()

array([ 4,  7,  6,  9, 15, 13, 20, 22, 19, 24, 30, 35, 39, 32, 33, 26, 44,
       43, 48, 37, 28, 17, 11,  0, 83, 70, 57, 46, 41, 52, 50, 63, 54,  2],
      dtype=int64)

In [32]:
df.head(2)

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0  1/1/2012 0:00    -1.8              -3.9         86                4   
1  1/1/2012 1:00    -1.8              -3.7         87                4   

   Visibility_km  Press_kPa Weather  
0            8.0     101.24     Fog  
1            8.0     101.24     Fog

#### Times when weather is exactly clear 

In [34]:
weather_clear=(df['Weather']=='Clear')

In [35]:
weather_clear

0       False
1       False
2       False
3       False
4       False
        ...  
8779    False
8780    False
8781    False
8782    False
8783    False
Name: Weather, Length: 8784, dtype: bool

In [36]:
weather_clear.value_counts()

Weather
False    7458
True     1326
Name: count, dtype: int64

In [37]:
weather_clear.describe()


count      8784
unique        2
top       False
freq       7458
Name: Weather, dtype: object

In [38]:
weather_clear.nunique()

2

##### lets print a dataframe when weather is clear

In [39]:
df[weather_clear]

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
67      1/3/2012 19:00   -16.9             -24.8         50               24   
114     1/5/2012 18:00    -7.1             -14.4         56               11   
115     1/5/2012 19:00    -9.2             -15.4         61                7   
116     1/5/2012 20:00    -9.8             -15.7         62                9   
117     1/5/2012 21:00    -9.0             -14.8         63               13   
...                ...     ...               ...        ...              ...   
8646   12/26/2012 6:00   -13.4             -14.8         89                4   
8698  12/28/2012 10:00    -6.1              -8.6         82               19   
8713   12/29/2012 1:00   -11.9             -13.6         87               11   
8714   12/29/2012 2:00   -11.8             -13.1         90               13   
8756  12/30/2012 20:00   -13.8             -16.5         80               24   

      Visibility_km  Press_kPa Weather  
67             25.0     101.74   Clear  
114            25.0     100.71   Clear  
115            25.0     100.80   Clear  
116            25.0     100.83   Clear  
117            25.0     100.83   Clear  
...             ...        ...     ...  
8646           25.0     102.47   Clear  
8698           24.1     101.27   Clear  
8713           25.0     101.31   Clear  
8714           25.0     101.33   Clear  
8756           25.0     101.52   Clear  

[1326 rows x 8 columns]

##### The third function that we can use is Groupby function. 
Here, we will create the group of all unique values of the Weather column and get the group of ‘Clear’ value 
only using the get_group function.
And, notice the answer is 1326 here.

In [41]:
df.groupby('Weather')

In [44]:
df.groupby('Weather').get_group('Clear')

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
67      1/3/2012 19:00   -16.9             -24.8         50               24   
114     1/5/2012 18:00    -7.1             -14.4         56               11   
115     1/5/2012 19:00    -9.2             -15.4         61                7   
116     1/5/2012 20:00    -9.8             -15.7         62                9   
117     1/5/2012 21:00    -9.0             -14.8         63               13   
...                ...     ...               ...        ...              ...   
8646   12/26/2012 6:00   -13.4             -14.8         89                4   
8698  12/28/2012 10:00    -6.1              -8.6         82               19   
8713   12/29/2012 1:00   -11.9             -13.6         87               11   
8714   12/29/2012 2:00   -11.8             -13.1         90               13   
8756  12/30/2012 20:00   -13.8             -16.5         80               24   

      Visibility_km  Press_kPa Weather  
67             25.0     101.74   Clear  
114            25.0     100.71   Clear  
115            25.0     100.80   Clear  
116            25.0     100.83   Clear  
117            25.0     100.83   Clear  
...             ...        ...     ...  
8646           25.0     102.47   Clear  
8698           24.1     101.27   Clear  
8713           25.0     101.31   Clear  
8714           25.0     101.33   Clear  
8756           25.0     101.52   Clear  

[1326 rows x 8 columns]

#### Let's find out when the weather was exaclty 4 km/h

In [51]:
wind_speed_4=(df.groupby('Wind Speed_km/h').get_group(4))

In [53]:
wind_speed_4

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0        1/1/2012 0:00    -1.8              -3.9         86                4   
1        1/1/2012 1:00    -1.8              -3.7         87                4   
96       1/5/2012 0:00    -8.8             -11.7         79                4   
101      1/5/2012 5:00    -7.0              -9.5         82                4   
146      1/7/2012 2:00    -8.1             -11.1         79                4   
...                ...     ...               ...        ...              ...   
8768   12/31/2012 8:00    -8.6             -10.3         87                4   
8769   12/31/2012 9:00    -8.1              -9.6         89                4   
8770  12/31/2012 10:00    -7.4              -8.9         89                4   
8772  12/31/2012 12:00    -5.8              -7.5         88                4   
8773  12/31/2012 13:00    -4.6              -6.6         86                4   

      Visibility_km  Press_kPa       Weather  
0               8.0     101.24           Fog  
1               8.0     101.24           Fog  
96              9.7     100.32          Snow  
101             4.0     100.19          Snow  
146            19.3     100.15        Cloudy  
...             ...        ...           ...  
8768            3.2     101.14  Snow Showers  
8769            2.4     101.09          Snow  
8770            6.4     101.05      Snow,Fog  
8772           12.9     100.78          Snow  
8773           12.9     100.63          Snow  

[474 rows x 8 columns]

In [54]:
wind_speed_4.value_counts()

Date/Time        Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  Visibility_km  Press_kPa  Weather      
1/1/2012 0:00    -1.8    -3.9              86         4                8.0            101.24     Fog              1
5/23/2012 4:00    12.6    9.4              81         4                24.1           100.87     Mostly Cloudy    1
5/5/2012 20:00    12.8   -2.0              36         4                25.0           101.62     Mainly Clear     1
5/5/2012 0:00     8.4     4.1              74         4                25.0           101.25     Mainly Clear     1
5/4/2012 18:00    17.4    13.0             75         4                24.1           100.79     Mostly Cloudy    1
                                                                                                                 ..
12/26/2012 8:00  -13.7   -15.2             88         4                16.1           102.53     Mostly Cloudy    1
12/26/2012 6:00  -13.4   -14.8             89         4                25.0  

In [55]:
wind_speed_4.describe()


Temp_C  Dew Point Temp_C   Rel Hum_%  Wind Speed_km/h  \
count  474.000000        474.000000  474.000000            474.0   
mean     7.022363          2.012869   72.833333              4.0   
std     11.397749         10.602057   16.791208              0.0   
min    -20.700000        -26.100000   28.000000              4.0   
25%     -1.700000         -6.575000   60.000000              4.0   
50%      7.550000          2.400000   76.000000              4.0   
75%     16.900000         10.900000   87.000000              4.0   
max     31.200000         22.000000  100.000000              4.0   

       Visibility_km   Press_kPa  
count     474.000000  474.000000  
mean       25.263080  101.438671  
std        11.791784    0.822204  
min         0.200000   99.110000  
25%        24.100000  100.922500  
50%        25.000000  101.475000  
75%        25.000000  101.967500  
max        48.300000  103.510000

In [56]:
wind_speed_4.shape

(474, 8)

#### Lets find out the missing value in the dataset
-To find the sum of Null values in a dataframe, we can use two functions i.e., isnull( ) or notnull( ).

In [57]:
df.isnull()

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0         False   False             False      False            False   
1         False   False             False      False            False   
2         False   False             False      False            False   
3         False   False             False      False            False   
4         False   False             False      False            False   
...         ...     ...               ...        ...              ...   
8779      False   False             False      False            False   
8780      False   False             False      False            False   
8781      False   False             False      False            False   
8782      False   False             False      False            False   
8783      False   False             False      False            False   

      Visibility_km  Press_kPa  Weather  
0             False      False    False  
1             False      False    False  
2             False      False    False  
3             False      False    False  
4             False      False    False  
...             ...        ...      ...  
8779          False      False    False  
8780          False      False    False  
8781          False      False    False  
8782          False      False    False  
8783          False      False    False  

[8784 rows x 8 columns]

In [58]:
df.isnull().sum()

Date/Time           0
Temp_C              0
Dew Point Temp_C    0
Rel Hum_%           0
Wind Speed_km/h     0
Visibility_km       0
Press_kPa           0
Weather             0
dtype: int64

In [59]:
#it means that there is no null value

In [60]:
#Alternatively, we can use notnull function, which is opposite of isnull( ).
df.notnull().sum()

Date/Time           8784
Temp_C              8784
Dew Point Temp_C    8784
Rel Hum_%           8784
Wind Speed_km/h     8784
Visibility_km       8784
Press_kPa           8784
Weather             8784
dtype: int64

#### Rename the column name ‘Weather’ of the dataframe to ‘Weather Condition’.

In [61]:
df.rename(columns={'Weather':'Weather_Conditions'})

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0        1/1/2012 0:00    -1.8              -3.9         86                4   
1        1/1/2012 1:00    -1.8              -3.7         87                4   
2        1/1/2012 2:00    -1.8              -3.4         89                7   
3        1/1/2012 3:00    -1.5              -3.2         88                6   
4        1/1/2012 4:00    -1.5              -3.3         88                7   
...                ...     ...               ...        ...              ...   
8779  12/31/2012 19:00     0.1              -2.7         81               30   
8780  12/31/2012 20:00     0.2              -2.4         83               24   
8781  12/31/2012 21:00    -0.5              -1.5         93               28   
8782  12/31/2012 22:00    -0.2              -1.8         89               28   
8783  12/31/2012 23:00     0.0              -2.1         86               30   

      Visibility_km  Press_kPa    Weather_Conditions  
0               8.0     101.24                   Fog  
1               8.0     101.24                   Fog  
2               4.0     101.26  Freezing Drizzle,Fog  
3               4.0     101.27  Freezing Drizzle,Fog  
4               4.8     101.23                   Fog  
...             ...        ...                   ...  
8779            9.7     100.13                  Snow  
8780            9.7     100.03                  Snow  
8781            4.8      99.95                  Snow  
8782            9.7      99.91                  Snow  
8783           11.3      99.89                  Snow  

[8784 rows x 8 columns]

Calculating the mean_visibility

In [62]:
df['Visibility_km'].mean()

27.664446721311478

In [63]:
df.Visibility_km.mean()

27.664446721311478

What is the Standard Deviation of ‘Pressure’ in this data?


In [64]:
df.Press_kPa.std()

0.8440047459486483

 What is the Variance of ‘Relative Humidity’ in this data ?

In [66]:
df['Rel Hum_%'].var()

286.24855019850196

##### Find all instances when ‘Snow’ was recorded.


In [70]:
weather_snow=(df.Weather=='Snow')

weather_snow.value_counts()

#### Find all instances when ‘Wind Speed is above 24’ and ‘Visibility is 25’.

In [81]:
find=(df['Wind Speed_km/h']>24) & (df['Visibility_km']==25)
find

0       False
1       False
2       False
3       False
4       False
        ...  
8779    False
8780    False
8781    False
8782    False
8783    False
Length: 8784, dtype: bool

In [76]:
find.value_counts()

Wind Speed_km/h
False    8784
Name: count, dtype: int64

In [82]:
df[find]

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
23      1/1/2012 23:00     5.3               2.0         79               30   
24       1/2/2012 0:00     5.2               1.5         77               35   
25       1/2/2012 1:00     4.6               0.0         72               39   
26       1/2/2012 2:00     3.9              -0.9         71               32   
27       1/2/2012 3:00     3.7              -1.5         69               33   
...                ...     ...               ...        ...              ...   
8705  12/28/2012 17:00    -8.6             -12.0         76               26   
8753  12/30/2012 17:00   -12.1             -15.8         74               28   
8755  12/30/2012 19:00   -13.4             -16.5         77               26   
8759  12/30/2012 23:00   -12.1             -15.1         78               28   
8760   12/31/2012 0:00   -11.1             -14.4         77               26   

      Visibility_km  Press_kPa        Weather  
23             25.0      99.31         Cloudy  
24             25.0      99.26   Rain Showers  
25             25.0      99.26         Cloudy  
26             25.0      99.26  Mostly Cloudy  
27             25.0      99.30  Mostly Cloudy  
...             ...        ...            ...  
8705           25.0     101.34   Mainly Clear  
8753           25.0     101.26   Mainly Clear  
8755           25.0     101.47   Mainly Clear  
8759           25.0     101.52  Mostly Cloudy  
8760           25.0     101.51         Cloudy  

[308 rows x 8 columns]

In [85]:
df[find].tail()

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
8705  12/28/2012 17:00    -8.6             -12.0         76               26   
8753  12/30/2012 17:00   -12.1             -15.8         74               28   
8755  12/30/2012 19:00   -13.4             -16.5         77               26   
8759  12/30/2012 23:00   -12.1             -15.1         78               28   
8760   12/31/2012 0:00   -11.1             -14.4         77               26   

      Visibility_km  Press_kPa        Weather  
8705           25.0     101.34   Mainly Clear  
8753           25.0     101.26   Mainly Clear  
8755           25.0     101.47   Mainly Clear  
8759           25.0     101.52  Mostly Cloudy  
8760           25.0     101.51         Cloudy

##### What is the Mean value of each column against each ‘Weather Condition’ ?

In [86]:
df.head()

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0  1/1/2012 0:00    -1.8              -3.9         86                4   
1  1/1/2012 1:00    -1.8              -3.7         87                4   
2  1/1/2012 2:00    -1.8              -3.4         89                7   
3  1/1/2012 3:00    -1.5              -3.2         88                6   
4  1/1/2012 4:00    -1.5              -3.3         88                7   

   Visibility_km  Press_kPa               Weather  
0            8.0     101.24                   Fog  
1            8.0     101.24                   Fog  
2            4.0     101.26  Freezing Drizzle,Fog  
3            4.0     101.27  Freezing Drizzle,Fog  
4            4.8     101.23                   Fog

In [87]:
df.groupby('Weather').mean('numeric only')

Temp_C  Dew Point Temp_C  \
Weather                                                                
Clear                                     6.825716          0.089367   
Cloudy                                    7.970544          2.375810   
Drizzle                                   7.353659          5.504878   
Drizzle,Fog                               8.067500          7.033750   
Drizzle,Ice Pellets,Fog                   0.400000         -0.700000   
Drizzle,Snow                              1.050000          0.150000   
Drizzle,Snow,Fog                          0.693333          0.120000   
Fog                                       4.303333          3.159333   
Freezing Drizzle                         -5.657143         -8.000000   
Freezing Drizzle,Fog                     -2.533333         -4.183333   
Freezing Drizzle,Haze                    -5.433333         -8.000000   
Freezing Drizzle,Snow                    -5.109091         -7.072727   
Freezing Fog                             -7.575000         -9.250000   
Freezing Rain                            -3.885714         -6.078571   
Freezing Rain,Fog                        -2.225000         -3.750000   
Freezing Rain,Haze                       -4.900000         -7.450000   
Freezing Rain,Ice Pellets,Fog            -2.600000         -3.700000   
Freezing Rain,Snow Grains                -5.000000         -7.300000   
Haze                                     -0.200000         -2.975000   
Mainly Clear                             12.558927          4.581671   
Moderate Rain,Fog                         1.700000          0.800000   
Moderate Snow                            -5.525000         -7.250000   
Moderate Snow,Blowing Snow               -5.450000         -6.500000   
Mostly Cloudy                            10.574287          3.131174   
Rain                                      9.786275          7.042810   
Rain Showers                             13.722340          9.187766   
Rain Showers,Fog                         12.800000         12.100000   
Rain Showers,Snow Showers                 2.150000         -1.500000   
Rain,Fog                                  8.273276          7.219828   
Rain,Haze                                 4.633333          2.066667   
Rain,Ice Pellets                          0.600000         -0.600000   
Rain,Snow                                 1.055556         -0.566667   
Rain,Snow Grains                          1.900000         -2.100000   
Rain,Snow,Fog                             0.800000          0.300000   
Rain,Snow,Ice Pellets                     1.100000         -0.175000   
Snow                                     -4.524103         -7.623333   
Snow Pellets                              0.700000         -6.400000   
Snow Showers                             -3.506667         -7.866667   
Snow Showers,Fog                        -10.675000        -11.900000   
Snow,Blowing Snow                        -5.410526         -7.621053   
Snow,Fog                                 -5.075676         -6.364865   
Snow,Haze                                -4.020000         -6.860000   
Snow,Ice Pellets                         -1.883333         -3.666667   
Thunderstorms                            24.150000         19.750000   
Thunderstorms,Heavy Rain Showers         10.900000          9.000000   
Thunderstorms,Moderate Rain Showers,Fog  19.600000         18.500000   
Thunderstorms,Rain                       20.433333         18.533333   
Thunderstorms,Rain Showers               20.037500         17.618750   
Thunderstorms,Rain Showers,Fog           21.600000         18.700000   
Thunderstorms,Rain,Fog                   20.600000         18.600000   

                                         Rel Hum_%  Wind Speed_km/h  \
Weather                                                               
Clear                                    64.497738        10.557315   
Cloudy                                   69.592593        16.127315   
Drizzle        

#### What is the Minimum & Maximum value of each column against each ‘Weather Condition’ ?



In [88]:
weather_condition=df.groupby('Weather').mean('numeric only')

In [93]:
weather_condition.mean()


Temp_C                3.303000
Dew Point Temp_C      0.802225
Rel Hum_%            84.599357
Wind Speed_km/h      17.851333
Visibility_km        10.518578
Press_kPa           100.467088
dtype: float64

In [94]:
df.groupby('Weather').min()

Date/Time  Temp_C  \
Weather                                                             
Clear                                      1/11/2012 1:00   -23.3   
Cloudy                                     1/1/2012 17:00   -21.4   
Drizzle                                   1/23/2012 21:00     1.1   
Drizzle,Fog                               1/23/2012 20:00     0.0   
Drizzle,Ice Pellets,Fog                   12/17/2012 9:00     0.4   
Drizzle,Snow                             12/17/2012 15:00     0.9   
Drizzle,Snow,Fog                         12/18/2012 21:00     0.3   
Fog                                         1/1/2012 0:00   -16.0   
Freezing Drizzle                          1/13/2012 10:00    -9.0   
Freezing Drizzle,Fog                        1/1/2012 2:00    -6.4   
Freezing Drizzle,Haze                      2/1/2012 11:00    -5.8   
Freezing Drizzle,Snow                      1/13/2012 3:00    -8.3   
Freezing Fog                               1/22/2012 6:00   -19.0   
Freezing Rain                             1/13/2012 11:00    -6.5   
Freezing Rain,Fog                         1/17/2012 23:00    -6.1   
Freezing Rain,Haze                         2/1/2012 14:00    -4.9   
Freezing Rain,Ice Pellets,Fog             12/17/2012 3:00    -2.6   
Freezing Rain,Snow Grains                  1/13/2012 9:00    -5.0   
Haze                                      1/22/2012 12:00   -11.5   
Mainly Clear                              1/10/2012 11:00   -22.8   
Moderate Rain,Fog                         12/10/2012 8:00     1.7   
Moderate Snow                             1/12/2012 15:00    -6.3   
Moderate Snow,Blowing Snow               12/27/2012 10:00    -5.5   
Mostly Cloudy                              1/1/2012 16:00   -23.2   
Rain                                       1/1/2012 18:00     0.3   
Rain Showers                               1/1/2012 22:00     1.6   
Rain Showers,Fog                          10/20/2012 3:00    12.8   
Rain Showers,Snow Showers                  11/4/2012 8:00     2.1   
Rain,Fog                                  1/23/2012 18:00     0.0   
Rain,Haze                                  3/13/2012 7:00     4.0   
Rain,Ice Pellets                          12/18/2012 5:00     0.6   
Rain,Snow                                  1/10/2012 5:00     0.6   
Rain,Snow Grains                          12/21/2012 0:00     1.9   
Rain,Snow,Fog                             12/8/2012 21:00     0.8   
Rain,Snow,Ice Pellets                     12/21/2012 1:00     0.9   
Snow                                       1/10/2012 1:00   -16.7   
Snow Pellets                             11/24/2012 15:00     0.7   
Snow Showers                               1/12/2012 7:00   -13.3   
Snow Showers,Fog                          12/26/2012 9:00   -11.3   
Snow,Blowing Snow                         1/13/2012 21:00   -12.0   
Snow,Fog                                 12/16/2012 15:00   -10.1   
Snow,Haze                                  2/1/2012 17:00    -4.3   
Snow,Ice Pellets                          12/10/2012 3:00    -4.3   
Thunderstorms                              7/16/2012 1:00    21.6   
Thunderstorms,Heavy Rain Showers           5/29/2012 6:00    10.9   
Thunderstorms,Moderate Rain Showers,Fog    7/17/2012 6:00    19.6   
Thunderstorms,Rain                        5/25/2012 20:00    19.4   
Thunderstorms,Rain Showers                5/29/2012 16:00    11.0   
Thunderstorms,Rain Showers,Fog             6/29/2012 3:00    19.5   
Thunderstorms,Rain,Fog                     7/17/2012 5:00    20.6   

                                         Dew Point Temp_C  Rel Hum_%  \
Weather                                                                
Clear                                               -28.5         20   
Cloudy                                              -26.8         18   
Drizzle                                              -0.2         74   
Drizzle,Fog                                          -1.6         85   
Drizzle,Ice Pellets,Fog    

when the weather was ‘Cloudy’ the minimum value of Temperature was -21.4, and when the weather was ‘Fog’ the minimum Pressure was 98.31.

In [96]:
df[df['Weather']=='Fog']

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0        1/1/2012 0:00    -1.8              -3.9         86                4   
1        1/1/2012 1:00    -1.8              -3.7         87                4   
4        1/1/2012 4:00    -1.5              -3.3         88                7   
5        1/1/2012 5:00    -1.4              -3.3         87                9   
6        1/1/2012 6:00    -1.5              -3.1         89                7   
...                ...     ...               ...        ...              ...   
8716   12/29/2012 4:00   -16.0             -17.2         90                6   
8717   12/29/2012 5:00   -14.8             -15.9         91                4   
8718   12/29/2012 6:00   -13.8             -15.3         88                4   
8719   12/29/2012 7:00   -14.8             -16.4         88                7   
8722  12/29/2012 10:00   -12.0             -13.3         90                7   

      Visibility_km  Press_kPa Weather  
0               8.0     101.24     Fog  
1               8.0     101.24     Fog  
4               4.8     101.23     Fog  
5               6.4     101.27     Fog  
6               6.4     101.29     Fog  
...             ...        ...     ...  
8716            9.7     101.25     Fog  
8717            6.4     101.25     Fog  
8718            9.7     101.25     Fog  
8719            8.0     101.22     Fog  
8722            6.4     101.15     Fog  

[150 rows x 8 columns]

In [98]:
#we can also do it by using groupby() function 
grouped=df.groupby('Weather')
fog_group=grouped.get_group('Fog')
fog_group

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0        1/1/2012 0:00    -1.8              -3.9         86                4   
1        1/1/2012 1:00    -1.8              -3.7         87                4   
4        1/1/2012 4:00    -1.5              -3.3         88                7   
5        1/1/2012 5:00    -1.4              -3.3         87                9   
6        1/1/2012 6:00    -1.5              -3.1         89                7   
...                ...     ...               ...        ...              ...   
8716   12/29/2012 4:00   -16.0             -17.2         90                6   
8717   12/29/2012 5:00   -14.8             -15.9         91                4   
8718   12/29/2012 6:00   -13.8             -15.3         88                4   
8719   12/29/2012 7:00   -14.8             -16.4         88                7   
8722  12/29/2012 10:00   -12.0             -13.3         90                7   

      Visibility_km  Press_kPa Weather  
0               8.0     101.24     Fog  
1               8.0     101.24     Fog  
4               4.8     101.23     Fog  
5               6.4     101.27     Fog  
6               6.4     101.29     Fog  
...             ...        ...     ...  
8716            9.7     101.25     Fog  
8717            6.4     101.25     Fog  
8718            9.7     101.25     Fog  
8719            8.0     101.22     Fog  
8722            6.4     101.15     Fog  

[150 rows x 8 columns]

#### Find all instances when ‘Weather is Clear’ or ‘Visibility is above 40’.

In [103]:
visible=(df['Weather']=='Clear')&(df['Visibility_km']>40)
df[visible]

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
351    1/15/2012 15:00   -15.4             -22.8         53               24   
352    1/15/2012 16:00   -15.1             -22.8         52               24   
425    1/18/2012 17:00   -11.3             -18.8         54               26   
440     1/19/2012 8:00   -13.7             -18.4         68               19   
441     1/19/2012 9:00   -12.7             -17.2         69               17   
...                ...     ...               ...        ...              ...   
8384   12/15/2012 8:00   -10.7             -15.6         67               13   
8385   12/15/2012 9:00   -10.4             -15.9         64               19   
8389  12/15/2012 13:00    -8.4             -14.7         60               19   
8631  12/25/2012 15:00    -7.1             -13.7         59               17   
8632  12/25/2012 16:00    -7.5             -13.9         60               11   

      Visibility_km  Press_kPa Weather  
351            48.3     102.71   Clear  
352            48.3     102.79   Clear  
425            48.3     101.54   Clear  
440            48.3     101.84   Clear  
441            48.3     101.73   Clear  
...             ...        ...     ...  
8384           48.3     102.69   Clear  
8385           48.3     102.74   Clear  
8389           48.3     102.64   Clear  
8631           48.3     101.98   Clear  
8632           48.3     102.03   Clear  

[313 rows x 8 columns]

In [ ]:
visible_pro=(df['Weather']=='Clear')&(df['Visibility_km']>40)&(df['